R/rdeephaven/R/client_wrapper.R

```diff
@@ -1,110 +1,249 @@
-      .verify_string("script", script)
-      private$internal_client$run_script(script)
     }
-  ),
-  private = list(
-    internal_client = NULL,
-    check_for_table = function(name) {
-      return(private$internal_client$check_for_table(name))
-    },
-    rbr_to_dh_table = function(rbr) {
-      ptr <- private$internal_client$new_arrow_array_stream_ptr()
-      rbr$export_to_c(ptr)
-      return(private$internal_client$new_table_from_arrow_array_stream_ptr(ptr))
-    },
-    arrow_to_dh_table = function(arrow_tbl) {
-      rbr <- as_record_batch_reader(arrow_tbl)
-      return(private$rbr_to_dh_table(rbr))
-    },
-    tibble_to_dh_table = function(tibbl) {
-      arrow_tbl <- arrow_table(tibbl)
-      return(private$arrow_to_dh_table(arrow_tbl))
-    },
-    df_to_dh_table = function(data_frame) {
-      arrow_tbl <- arrow_table(data_frame)
-      return(private$arrow_to_dh_table(arrow_tbl))
+
+    if (int_option != "") {
+      verify_string("int_option", int_option, TRUE)
+      new_int_option <- strsplit(int_option, ":", fixed = TRUE)
+      options$add_int_option(new_int_option[1], as.numeric(new_int_option[2]))
     }
-  )
+
+    if (string_option != "") {
+      verify_string("string_option", string_option, TRUE)
+      new_string_option <- strsplit(string_option, ":", fixed = TRUE)
+      options$add_string_option(new_string_option[1], new_string_option[2])
+    }
+
+    if (extra_header != "") {
+      verify_string("extra_header", extra_header, TRUE)
+      new_extra_header <- strsplit(extra_header, ":", fixed = TRUE)
+      options$add_extra_header(new_extra_header[1], new_extra_header[2])
+    }
+
+    internal_client <- new(INTERNAL_Client,
+      target = target,
+      client_options = options
+    )
+    return(new("Client", .internal_rcpp_object = internal_client))
+  }
+)
+
+### HELPER FUNCTIONS ###
+# These functions return RC objects returned by Rcpp without wrapping them in S4
+
+check_for_table <- function(client, name) {
+  return(client@.internal_rcpp_object$check_for_table(name))
+}
+
+rbr_to_dh_table <- function(client, rbr) {
+  ptr <- client@.internal_rcpp_object$new_arrow_array_stream_ptr()
+  rbr$export_to_c(ptr)
+  return(client@.internal_rcpp_object$new_table_from_arrow_array_stream_ptr(ptr))
+}
+
+arrow_to_dh_table <- function(client, arrow_tbl) {
+  rbr <- as_record_batch_reader(arrow_tbl)
+  return(rbr_to_dh_table(client, rbr))
+}
+
+tibble_to_dh_table <- function(client, tibbl) {
+  arrow_tbl <- arrow_table(tibbl)
+  return(arrow_to_dh_table(client, arrow_tbl))
+}
+
+df_to_dh_table <- function(client, data_frame) {
+  arrow_tbl <- arrow_table(data_frame)
+  return(arrow_to_dh_table(client, arrow_tbl))
+}
+
+### USER-FACING METHODS ###
+
+setGeneric(
+  "open_table",
+  function(client_instance, name) {
+    return(standardGeneric("open_table"))
+  },
+  signature = c("client_instance", "name")
+)
+
+#' @export
+setMethod(
+  "open_table",
+  signature = c(client_instance = "Client"),
```

Comment:
This comment applies everywhere.

The signature provided in the methods should be as specific as possible to avoid crappy dispatch.  In this case, the function takes a `Client` and a string.  As such, the signature should indicate both.  Just defaulting to single dispatch is bad practice and is the whole reason S4 was created.

Reply:
My only argument against this is that the error messages produced by R being unable to find a method to dispatch are not very helpful. So, if you use S4 dispatch to insist that `name` is a string, you have no control over the error message it gives off, and the one it gives off is not as informative as the one produced by `verify_string`.